packages/backend-common/src/reading/AzureBlobStorageUrlReader.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  ReaderFactory,
+  ReadTreeResponse,
+  ReadUrlOptions,
+  ReadUrlResponse,
+  SearchResponse,
+  UrlReader,
+} from './types';
+import {
+  AzureBlobStorageIntegration,
+  ScmIntegrations,
+} from '@backstage/integration';
+import { ReadUrlResponseFactory } from './ReadUrlResponseFactory';
+import { NotFoundError, NotModifiedError } from '@backstage/errors';
+import { parseLastModified } from './util';
+import fetch, { Response } from 'node-fetch';
+
+/**
+ * Implements a {@link @backstage/backend-plugin-api#UrlReaderService} for Azure BlobStorage Container.
+ *
+ * @public
+ */
+export class AzureBlobStorageUrlReader implements UrlReader {
+  static factory: ReaderFactory = ({ config }) => {
+    const integrations = ScmIntegrations.fromConfig(config);
+    return integrations.azureBlobStorage.list().map(integration => {
+      const reader = new AzureBlobStorageUrlReader(integration);
+      const predicate = (url: URL) => url.host === integration.config.host;
+      return { reader, predicate };
+    });
+  };
+
+  constructor(private readonly integration: AzureBlobStorageIntegration) {}
+
+  async readUrl(
+    url: string,
+    options?: ReadUrlOptions,
+  ): Promise<ReadUrlResponse> {
+    const { etag, lastModifiedAfter } = options ?? {};
+    let response: Response;
+    try {
+      response = await fetch(
+        `${url}${this.integration.config.secretAccessKey}`,
+        {
+          headers: {
+            ...(etag && { 'If-None-Match': etag }),
+            ...(lastModifiedAfter && {
+              'If-Modified-Since': lastModifiedAfter.toUTCString(),
+            }),
+          },
+        },
+      );
+    } catch (e) {
+      throw new Error(`Unable to read ${url}, ${e}`);
```

Comment:
Given the signature is included in the querysting, is there any risk of leaking secrets from the fetch api's error?  (e.g. `Failed to read from https://foo.blob.core.windows.net/path/to/file.json?SECRET` )